Segmentation fault on docker run after pulling remote image

<!--
If you are reporting a new issue, make sure that we do not have any duplicates
already open. You can ensure this by searching the issue list for this
repository. If there is a duplicate, please close your issue and add a comment
to the existing issue instead.

If you suspect your issue is a bug, please edit your issue description to
include the BUG REPORT INFORMATION shown below. If you fail to provide this
information within 7 days, we cannot debug your issue and will close it. We
will, however, reopen it if you later provide the information.

For more information about reporting issues, see
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#reporting-other-issues

---------------------------------------------------
GENERAL SUPPORT INFORMATION
---------------------------------------------------

The GitHub issue tracker is for bug reports and feature requests.
General support can be found at the following locations:

- Docker Support Forums - https://forums.docker.com
- IRC - irc.freenode.net #docker channel
- Post a question on StackOverflow, using the Docker tag

---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------
Use the commands below to provide key information from your environment:
You do NOT have to include this information if this is a FEATURE REQUEST
-->

**Description**

<!--
Briefly describe the problem you are having in a few paragraphs.
-->

**Steps to reproduce the issue:**
<details>
<summary> Original issue </summary>

1. Build a package using the docker/cli@e636a538 and moby/moby@184cea5ff
2. Install package on ubuntu xenial
3. Run `docker run -d --name server --net testbind -p 8080:8080/udp appropriate/nc sh -c 'while true; do echo hello | nc -w 1 -lu 8080; done'`

</details>

1. Build a package using the docker/cli@e636a538 and moby/moby@184cea5ff
2. Install package on ubuntu xenial
3. Run `docker run <any_remote_image> echo hello`

EDIT:
It appears this is occurring after pulling **any** images from a remote repository

**Describe the results you received:**
<details>
<summary> Original issue </summary>

```
root@ubuntu-xenial:/vagrant# docker run -d --name server --net testbind -p 8080:8080/udp appropriate/nc sh -c 'while true; do echo hello | nc -w 1 -lu 8080; done'
Unable to find image 'appropriate/nc:latest' locally
latest: Pulling from appropriate/nc
12b41071e6ce: Pull complete
d41ebf7a19c0: Pull complete
0b849cb5c85c: Pull complete
a3ed95caeb02: Pull complete
Digest: sha256:ff0f2b54fb1e1bfe277047fd660415310232d301e27bec9423aa741de961e6d4
Status: Downloaded newer image for appropriate/nc:latest
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa8b605]

goroutine 1 [running]:
github.com/docker/cli/cli/command/container.runContainer(0xc4201cf740, 0xc4203a8150, 0xc420365400, 0xc420403ec0, 0x0, 0x0)
	/go/src/github.com/docker/cli/cli/command/container/run.go:170 +0xb75
github.com/docker/cli/cli/command/container.runRun(0xc4201cf740, 0xc42007d4d0, 0xc4203a8150, 0xc420365400, 0x0, 0x0)
	/go/src/github.com/docker/cli/cli/command/container/run.go:117 +0x39b
github.com/docker/cli/cli/command/container.NewRunCommand.func1(0xc420399d40, 0xc4203749a0, 0x4, 0xb, 0x0, 0x0)
	/go/src/github.com/docker/cli/cli/command/container/run.go:49 +0x109
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).execute(0xc420399d40, 0xc4200101c0, 0xb, 0xb, 0xc420399d40, 0xc4200101c0)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:646 +0x44e
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc420366480, 0xc420366900, 0xc420331f10, 0xc42000e4a8)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:742 +0x349
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).Execute(0xc420366480, 0xc420366480, 0x161a780)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:695 +0x2b
main.main()
	/go/src/github.com/docker/cli/cmd/docker/docker.go:168 +0xc3
```

</details>

```
root@ubuntu-xenial:/vagrant# docker run centos:7 echo hello
Unable to find image 'centos:7' locally
7: Pulling from library/centos
74f0853ba93b: Pull complete
Digest: sha256:26f74cefad82967f97f3eeeef88c1b6262f9b42bc96f2ad61d6f3fdf544759b8
Status: Downloaded newer image for centos:7
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa8b605]

goroutine 1 [running]:
github.com/docker/cli/cli/command/container.runContainer(0xc4201cf740, 0xc4203aa150, 0xc420365400, 0xc420403de0, 0x0, 0x0)
	/go/src/github.com/docker/cli/cli/command/container/run.go:170 +0xb75
github.com/docker/cli/cli/command/container.runRun(0xc4201cf740, 0xc42007d4d0, 0xc4203aa150, 0xc420365400, 0x0, 0x0)
	/go/src/github.com/docker/cli/cli/command/container/run.go:117 +0x39b
github.com/docker/cli/cli/command/container.NewRunCommand.func1(0xc420397d40, 0xc420405650, 0x3, 0x3, 0x0, 0x0)
	/go/src/github.com/docker/cli/cli/command/container/run.go:49 +0x109
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).execute(0xc420397d40, 0xc4200100c0, 0x3, 0x3, 0xc420397d40, 0xc4200100c0)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:646 +0x44e
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc420366480, 0xc420366900, 0xc420331f10, 0xc42000e4a8)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:742 +0x349
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).Execute(0xc420366480, 0xc420366480, 0x161a780)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:695 +0x2b
main.main()
	/go/src/github.com/docker/cli/cmd/docker/docker.go:168 +0xc3
```
**Describe the results you expected:**
<details>
<summary> Original issue </summary>

```
root@ubuntu-xenial:/home/ubuntu# docker run -d --name server --net testbind -p 8080:8080/udp appropriate/nc sh -c 'while true; do echo hello | nc -w 1 -lu 8080; done'
Unable to find image 'appropriate/nc:latest' locally
latest: Pulling from appropriate/nc
12b41071e6ce: Pull complete
d41ebf7a19c0: Pull complete
0b849cb5c85c: Pull complete
a3ed95caeb02: Pull complete
Digest: sha256:ff0f2b54fb1e1bfe277047fd660415310232d301e27bec9423aa741de961e6d4
113e0258b1defc42de45dba364b664eb861ca2b9351aab677aa72279f429a94b
```

</details>

```
root@ubuntu-xenial:/vagrant# docker run centos:7 echo hello
hello
```

**Additional information you deem important (e.g. issue happens only occasionally):**
Issue seems to disappear after remote image has been pulled.

```
root@ubuntu-xenial:/vagrant# docker run centos:7 echo hello
Unable to find image 'centos:7' locally
7: Pulling from library/centos
74f0853ba93b: Pull complete
Digest: sha256:26f74cefad82967f97f3eeeef88c1b6262f9b42bc96f2ad61d6f3fdf544759b8
Status: Downloaded newer image for centos:7
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa8b605]

goroutine 1 [running]:
github.com/docker/cli/cli/command/container.runContainer(0xc4201cf740, 0xc4203aa150, 0xc420365400, 0xc420403de0, 0x0, 0x0)
	/go/src/github.com/docker/cli/cli/command/container/run.go:170 +0xb75
github.com/docker/cli/cli/command/container.runRun(0xc4201cf740, 0xc42007d4d0, 0xc4203aa150, 0xc420365400, 0x0, 0x0)
	/go/src/github.com/docker/cli/cli/command/container/run.go:117 +0x39b
github.com/docker/cli/cli/command/container.NewRunCommand.func1(0xc420397d40, 0xc420405650, 0x3, 0x3, 0x0, 0x0)
	/go/src/github.com/docker/cli/cli/command/container/run.go:49 +0x109
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).execute(0xc420397d40, 0xc4200100c0, 0x3, 0x3, 0xc420397d40, 0xc4200100c0)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:646 +0x44e
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc420366480, 0xc420366900, 0xc420331f10, 0xc42000e4a8)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:742 +0x349
github.com/docker/cli/vendor/github.com/spf13/cobra.(*Command).Execute(0xc420366480, 0xc420366480, 0x161a780)
	/go/src/github.com/docker/cli/vendor/github.com/spf13/cobra/command.go:695 +0x2b
main.main()
	/go/src/github.com/docker/cli/cmd/docker/docker.go:168 +0xc3
root@ubuntu-xenial:/vagrant# docker run centos:7 echo hello
hello
```

<details>
<summary> Original issue </summary>

Issue is consistent and is included in the test:


https://github.com/moby/moby/blob/184cea5ff710abde25547749e5608b24a255ba09/integration-cli/docker_cli_network_unix_test.go#L1812-L1822

</details>

**Output of `docker version`:**
Don't mind the `17.06.0-dev` it's from the base [moby/moby/VERSION](https://github.com/moby/moby/blob/master/VERSION) file

```
root@ubuntu-xenial:/vagrant# docker version
Client:
 Version:      17.06.0-dev
 API version:  1.31
 Go version:   go1.8.3
 Git commit:   184cea5ff
 Built:        Thu Aug 31 17:53:14 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-dev
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   184cea5ff
 Built:        Thu Aug 31 17:52:09 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
root@ubuntu-xenial:/vagrant# docker info
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 17.06.0-dev
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-87-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 992.2MiB
Name: ubuntu-xenial
ID: VPBN:G2KS:AESQ:RVVR:3WCT:PQL3:GRN3:QBD7:INWL:TUYA:54ZS:2TVB
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
This is running in a Vagrant box, system information:
```
root@ubuntu-xenial:/vagrant# cat /etc/os-release
NAME="Ubuntu"
VERSION="16.04.3 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.3 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```

--- a/cli/command/container/client_test.go
+++ b/cli/command/container/client_test.go
@@ -1,16 +1,23 @@
 package container
 
 import (
+	"io"
+
 	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
 	"github.com/docker/docker/client"
 	"golang.org/x/net/context"
 )
 
 type fakeClient struct {
 	client.Client
-	inspectFunc     func(string) (types.ContainerJSON, error)
-	execInspectFunc func(execID string) (types.ContainerExecInspect, error)
-	execCreateFunc  func(container string, config types.ExecConfig) (types.IDResponse, error)
+	inspectFunc         func(string) (types.ContainerJSON, error)
+	execInspectFunc     func(execID string) (types.ContainerExecInspect, error)
+	execCreateFunc      func(container string, config types.ExecConfig) (types.IDResponse, error)
+	createContainerFunc func(config *container.Config, hostConfig *container.HostConfig, networkingConfig *network.NetworkingConfig, containerName string) (container.ContainerCreateCreatedBody, error)
+	imageCreateFunc     func(parentReference string, options types.ImageCreateOptions) (io.ReadCloser, error)
+	infoFunc            func() (types.Info, error)
 }
 
 func (f *fakeClient) ContainerInspect(_ context.Context, containerID string) (types.ContainerJSON, error) {
@@ -36,4 +43,31 @@ func (f *fakeClient) ContainerExecInspect(_ context.Context, execID string) (typ
 
 func (f *fakeClient) ContainerExecStart(ctx context.Context, execID string, config types.ExecStartCheck) error {
 	return nil
+}
+
+func (f *fakeClient) ContainerCreate(
+	_ context.Context,
+	config *container.Config,
+	hostConfig *container.HostConfig,
+	networkingConfig *network.NetworkingConfig,
+	containerName string,
+) (container.ContainerCreateCreatedBody, error) {
+	if f.createContainerFunc != nil {
+		return f.createContainerFunc(config, hostConfig, networkingConfig, containerName)
+	}
+	return container.ContainerCreateCreatedBody{}, nil
+}
+
+func (f *fakeClient) ImageCreate(ctx context.Context, parentReference string, options types.ImageCreateOptions) (io.ReadCloser, error) {
+	if f.imageCreateFunc != nil {
+		return f.imageCreateFunc(parentReference, options)
+	}
+	return nil, nil
+}
+
+func (f *fakeClient) Info(_ context.Context) (types.Info, error) {
+	if f.infoFunc != nil {
+		return f.infoFunc()
+	}
+	return types.Info{}, nil
 }

--- a/cli/command/container/create.go
+++ b/cli/command/container/create.go
@@ -198,7 +198,7 @@ func createContainer(ctx context.Context, dockerCli command.Cli, containerConfig
 			fmt.Fprintf(stderr, "Unable to find image '%s' locally\n", reference.FamiliarString(namedRef))
 
 			// we don't want to write to stdout anything apart from container.ID
-			if err = pullImage(ctx, dockerCli, config.Image, stderr); err != nil {
+			if err := pullImage(ctx, dockerCli, config.Image, stderr); err != nil {
 				return nil, err
 			}
 			if taggedRef, ok := namedRef.(reference.NamedTagged); ok && trustedRef != nil {
@@ -212,8 +212,9 @@ func createContainer(ctx context.Context, dockerCli command.Cli, containerConfig
 			if retryErr != nil {
 				return nil, retryErr
 			}
+		} else {
+			return nil, err
 		}
-		return nil, err
 	}
 
 	for _, warning := range response.Warnings {

--- a/cli/command/container/create_test.go
+++ b/cli/command/container/create_test.go
@@ -1,13 +1,20 @@
 package container
 
 import (
+	"context"
+	"io"
+	"io/ioutil"
 	"os"
+	"strings"
 	"testing"
 
-	"io/ioutil"
-
+	"github.com/docker/cli/internal/test"
 	"github.com/docker/cli/internal/test/testutil"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
 	"github.com/gotestyourself/gotestyourself/fs"
+	"github.com/pkg/errors"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
@@ -62,3 +69,52 @@ func TestCIDFileCloseWithWrite(t *testing.T) {
 	_, err = os.Stat(path)
 	require.NoError(t, err)
 }
+
+func TestCreateContainerPullsImageIfMissing(t *testing.T) {
+	imageName := "does-not-exist-locally"
+	responseCounter := 0
+	containerID := "abcdef"
+
+	client := &fakeClient{
+		createContainerFunc: func(
+			config *container.Config,
+			hostConfig *container.HostConfig,
+			networkingConfig *network.NetworkingConfig,
+			containerName string,
+		) (container.ContainerCreateCreatedBody, error) {
+			defer func() { responseCounter++ }()
+			switch responseCounter {
+			case 0:
+				return container.ContainerCreateCreatedBody{}, fakeNotFound{}
+			case 1:
+				return container.ContainerCreateCreatedBody{ID: containerID}, nil
+			default:
+				return container.ContainerCreateCreatedBody{}, errors.New("unexpected")
+			}
+		},
+		imageCreateFunc: func(parentReference string, options types.ImageCreateOptions) (io.ReadCloser, error) {
+			return ioutil.NopCloser(strings.NewReader("")), nil
+		},
+		infoFunc: func() (types.Info, error) {
+			return types.Info{IndexServerAddress: "http://indexserver"}, nil
+		},
+	}
+	cli := test.NewFakeCli(client)
+	config := &containerConfig{
+		Config: &container.Config{
+			Image: imageName,
+		},
+		HostConfig: &container.HostConfig{},
+	}
+	body, err := createContainer(context.Background(), cli, config, "name")
+	require.NoError(t, err)
+	expected := container.ContainerCreateCreatedBody{ID: containerID}
+	assert.Equal(t, expected, *body)
+	stderr := cli.ErrBuffer().String()
+	assert.Contains(t, stderr, "Unable to find image 'does-not-exist-locally:latest' locally")
+}
+
+type fakeNotFound struct{}
+
+func (f fakeNotFound) NotFound() bool { return true }
+func (f fakeNotFound) Error() string  { return "error fake not found" }

--- a/cli/command/trust.go
+++ b/cli/command/trust.go
@@ -12,6 +12,10 @@ var (
 	untrusted bool
 )
 
+func init() {
+	untrusted = !getDefaultTrustState()
+}
+
 // AddTrustVerificationFlags adds content trust flags to the provided flagset
 func AddTrustVerificationFlags(fs *pflag.FlagSet) {
 	trusted := getDefaultTrustState()

